### PR TITLE
Bug 1913960: rebase on top of kubernetes/autoscaler 1.20

### PR DESCRIPTION
**Pull request by elmiko**
# 1.20 autoscaler rebase process

inspired by the commit description for the 1.19 rebase.
[pr #164](https://github.com/openshift/kubernetes-autoscaler/pull/164)

identify carry commits:
```
git log --oneline --no-merges 78d401fd83e3a003c01056fff3f955850d06a756..openshift/master
```
where `78d401` reflects the changes since our last rebase (1.19.0). this is the
list of commits we will need to apply onto the new upstream version of the
autoscaler. ideally, some of these commits can be dropped.

After identifying the carry commits, the next step is to create the new commit-tree that
will be used for the rebase and then cherry pick the carry commits into the new branch.
The following commands cover these steps:

**Process**

```
$ git remote update # make sure we update our refs
$ git checkout 57cf6d00ef8af2760f3d94dc789002ef92eb4689 # cluster-autoscaler-1.20
$ git checkout -b merge-tmp # create a branch to do our merge work from
$ git checkout openshift/master # we want to be at the tip of the openshift master branch when we run the next command
$ echo 'merge upstream/cluster-autoscaler-1.20' | git commit-tree merge-tmp^{tree} -p HEAD -p merge-tmp -F -
deadbeef12345678 # id of new merge commit
$ git branch merge-1.20 deadbeef12345678 # create a new branch for the cherry-pick work
$ git checkout merge-1.20 # make sure we are on the proper branch
$ git cherry-pick <carry commits>
```

With the `merge-1.20` branch in place, I cherry picked the carry commits which applied, resolved merge conflicts,
and finally tested the resulting tree against the unit test and end-to-end suite.

**Carried Commits**

These commits are for features which have not yet been accepted upstream, are integral to our CI platform, or are
specific to the releases we create for OpenShift.

```
ac7890394 UPSTREAM: <carry>: openshift: add fixes for unstructured scalable resources
3ad14d355 UPSTREAM: 3708: openshift: Bump github.com/heketi/heketi to v10.1.0
a4099f2d7 UPSTREAM: <carry>: Updating vertical-pod-autoscaler builder & base images to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/f82a216a6a3707b80a635bace9367f1a8288b7a7/images/vertical-pod-autoscaler.yml
30a061c91 UPSTREAM: <carry>: Updating atomic-openshift-cluster-autoscaler builder & base images to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/f82a216a6a3707b80a635bace9367f1a8288b7a7/images/atomic-openshift-cluster-autoscaler.yml
22838a30c UPSTREAM: <carry>: openshift: Add tests for nodegroup TemplateNodeInfo
5b045be8c UPSTREAM: <carry>: openshift: Copy standard labels from an existing node if one exists
889ce351a UPSTREAM: <carry>: openshift: Ensure stable versions of generic labels are also included
55c138cd5 UPSTREAM: <carry>: openshift: Updating vertical-pod-autoscaler/Dockerfile.rhel baseimages to match ocp-build-data config
c0ac04d2b UPSTREAM: <carry>: openshift: Updating images/cluster-autoscaler/Dockerfile.rhel baseimages to match ocp-build-data config
3fabddac9 UPSTREAM: <carry>: openshift: remove Dockerfile.rhel7 files for ca and vpa
31d20311d UPSTREAM: <carry>: openshift: Add vertical-pod-autoscaler/Dockerfile.rhel to match build configuration in ocp-build-data
ac7669062 UPSTREAM: <carry>: openshift: Add images/cluster-autoscaler/Dockerfile.rhel to match build configuration in ocp-build-data
88e10aa44 UPSTREAM: <carry>: openshift: Update VPA builder images to use go 1.14
bce8208b3 UPSTREAM: <carry>: openshift: Implement scale from zero
a2156bf25 UPSTREAM: <carry>: openshift: Rename FailureMessage to ErrorMessage
45a0bd5e3 UPSTREAM: <carry>: openshift: update cluster-autoscaler OWNERS
f8624cb76 UPSTREAM: <carry>: openshift: Add OpenShift VPA image builds
104f3d494 UPSTREAM: <carry>: openshift: Revert "Adding config for event filtering"
d008d774a UPSTREAM: <carry>: openshift: Set default annotations, and cluster name label to use machine.openshift.io
937d51963 UPSTREAM: <carry>: Bump scripts go version to 1.13 and update path to cloudprovider/clusterapi
17de84708 UPSTREAM: <carry>: openshift: Extend makefile with 'make goimports' target
1a1f1fc76 UPSTREAM: <carry>: Fix git commit message verification script
9dd65caae UPSTREAM: <carry>: openshift: Switch builds to use Go 1.12
0ba73333a UPSTREAM: <carry>: openshift: create git history verification script
a2760f607 UPSTREAM: <carry>: openshift: Add fmt, lint, vet scripts/Makefile
0f70d6ae2 UPSTREAM: <carry>: openshift: Add a RHEL7 dockerfile and standarize format
a8e852169 UPSTREAM: <carry>: openshift: cluster-autoscaler.spec: set golang_version to 1.10
ce4bdc21d UPSTREAM: <carry>: openshift: cluster-autoscaler.spec: bump golang_version to 1.10.4
906f8f7eb UPSTREAM: <carry>: openshift: Fix spec file to be consistent
eafb0c97b UPSTREAM: <carry>: openshift: Bump embedded tools
99ec67a06 UPSTREAM: <carry>: openshift: Fix the spec and hack scripts so the package can be built in both build systems
5381d0447 UPSTREAM: <carry>: openshift: Add openshift/release Makefile and hack scripts
381e5a654 UPSTREAM: <carry>: openshift: Add dockerfile for cluster autoscaler.
7665c1f10 UPSTREAM: <carry>: openshift: Add spec file for cluster-autoscaler.
```

**Dropped Commits**

These commits were carried over the 1.19.x release history and represent work that has been accepted upstream.

```
99572119c UPSTREAM: <drop>: openshift: Update group identifier to use for Cluster API annotations
61b8048c4 UPSTREAM: <drop>: openshift: Support using --cloud-config for clusterapi provider
bf74515dc UPSTREAM: <carry>: openshift: add fixes for node autodiscovery
71ef9a5ba UPSTREAM: <drop>: openshift: Add node autodiscovery to cluster-autoscaler clusterapi provider
08af78fb5 UPSTREAM: <drop>: openshift: Improve Cluster API tests to work better with constrained resources
08111cefe UPSTREAM: <drop>: openshift: Convert clusterapi provider to use unstructured
19a57d277 UPSTREAM: <drop>: openshift: Update vendor to pull in necessary new paths for client-go
adaf35c01 UPSTREAM: <carry>: openshift: Fallback to Status Replicas if Replicas nil when listing NodeGroups
bb9b9bbe2 UPSTREAM: <carry>: openshift: Fallback to status if replicas nil in spec
```

### CODE REVIEW

File: cluster-autoscaler/Dockerfile [hunk 1]
```diff
@@ -1,19 +0,0 @@
-# Copyright 2016 The Kubernetes Authors. All rights reserved
```
Comment (JoelSpeed) [inline cluster-autoscaler/Dockerfile:1]: How does our release image get built? Is this going to affect that?

**Review by JoelSpeed — COMMENTED** [review]
We should check that the new workflows that have been added won't affect our PRs

Changes in the autoscaler to be aware of:
- Daemonset Pods are now evicted on drain, but no error is returned if this fails

File: cluster-autoscaler/Dockerfile (hunk 1 above)
Comment (JoelSpeed) [reply cluster-autoscaler/Dockerfile:1]: We have our own dockerfiles in images/cluster-autoscaler, this doesn't affect our build process, no issue here

Comment (elmiko) [reply cluster-autoscaler/Dockerfile:1]: we should exclusively be using the `./images/cluster-autoscaler/Dockerfile.rhel` file to build our stuff.

**Pull request merged**